core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java

```diff
@@ -328,13 +330,24 @@ protected int getNewForOldInput(int oldOrdinal) {
     assert newInput != null;
 
     RelDataType oldInputType = oldInput.getRowType();
-    final int newOffset = calculateFlattenedOffset(oldInputType, oldOrdinal);
+    final int newOffset = calculateFlattenedOffset(oldInputType, oldOrdinal) + existingOffset;
     newOrdinal += newOffset;
     final RelDataTypeField field =
         newInput.getRowType().getFieldList().get(newOffset);
     return Ord.of(newOrdinal, field.getType());
   }
 
+  /**
+   * Maps the ordinal of a field pre-flattening to the ordinal of the
+   * corresponding field post-flattening, and also returns its type.
+   *
+   * @param oldOrdinal Pre-flattening ordinal
+   * @return Post-flattening ordinal and type
+   */
+  protected Ord<RelDataType> getNewFieldForOldInput(int oldOrdinal) {
```

Comment:
@zabetak This is the old function signature so I have to keep them to keep backward compatibility.